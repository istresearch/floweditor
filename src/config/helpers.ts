import {
  FlowTypes,
  FlowTypeVisibility,
  Operator,
  Type,
  VISIBILITY_ONLINE
} from 'config/interfaces';
import { FlowEditorConfig } from 'flowTypes';

export const isOnlineFlowType = (flowType: FlowTypes) => {
  return !!VISIBILITY_ONLINE.find((type: FlowTypes) => type === flowType);
};

export const filterOperators = (operators: Operator[], config: FlowEditorConfig): Operator[] => {
  return filterVisibility(operators, config);
};

export const filterTypeConfigs = (typeConfigs: Type[], config: FlowEditorConfig): Type[] => {
  return filterVisibility(typeConfigs, config);
};

const filterVisibility = (items: FlowTypeVisibility[], config: FlowEditorConfig): any[] => {
  return items.filter((item: FlowTypeVisibility) => {
    // if we have a filter on our type, don't return it unless its present in our config
    if (item.filter) {
      if (!(config.filters || []).find((name: string) => name === item.filter)) {
        return false;
      }
    }

    if (item.visibility === undefined) {
      return true;
    }

    return item.visibility.findIndex((ft: FlowTypes) => ft === config.flowType) > -1;
  });
};
