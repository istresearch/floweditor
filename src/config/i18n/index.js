import i18n from 'i18next';
import { initReactI18next } from 'react-i18next';
import LanguageDetector from 'i18next-browser-languagedetector';

// TODO: consider chunking out translation files
// and using something like XHR backend to lazily
// load only the required language.
// See https://www.i18next.com/how-to/add-or-load-translations

import defaults from './defaults.json';
import cs from './cs/resource.json';
import es from './es/resource.json';
import fr from './fr/resource.json';
import mn from './mn/resource.json';
import ptBR from './pt-br/resource.json';

i18n
  .use(LanguageDetector)
  .use(initReactI18next)
  .init({
    detection: {
      order: ['querystring', 'htmlTag']
    },
    resources: {
      defaults: { translation: defaults },
      cs: { translation: cs },
      es: { translation: es },
      fr: { translation: fr },
      mn: { translation: mn },
      'pt-br': { translation: ptBR }
    },
    fallbackLng: ['defaults'],
    interpolation: {
      prefix: '[[',
      suffix: ']]'
    },
    lowerCaseLng: true
  });

export default i18n;
