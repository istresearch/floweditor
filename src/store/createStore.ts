import { applyMiddleware, createStore, Middleware } from 'redux';
import { composeWithDevTools } from '@redux-devtools/extension';
import thunk from 'redux-thunk';

import rootReducer from 'store/rootReducer';
import AppState, { initialState } from 'store/state';

const middlewares: Middleware[] = [thunk];

export default (state: AppState = initialState) => {
  const store = createStore(
    rootReducer,
    state,
    composeWithDevTools(applyMiddleware(...middlewares))
  );

  /* istanbul ignore next */
  if (module.hot) {
    module.hot.accept('./rootReducer', () => {
      const { default: nextRootReducer } = require('./rootReducer');
      store.replaceReducer(nextRootReducer);
    });
  }

  return store;
};
