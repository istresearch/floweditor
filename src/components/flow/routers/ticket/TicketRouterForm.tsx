import { default as bindCallbacks } from 'auto-bind/react';
import Dialog, { ButtonSet } from 'components/dialog/Dialog';
import { hasErrors, renderIssues } from 'components/flow/actions/helpers';
import { RouterFormProps } from 'components/flow/props';
import { nodeToState, stateToNode } from './helpers';
import { createResultNameInput } from 'components/flow/routers/widgets';
import TypeList from 'components/nodeeditor/TypeList';
import * as React from 'react';
import { FormState, mergeForm, StringEntry, FormEntry } from 'store/nodeEditor';
import {
  Alphanumeric,
  Required,
  shouldRequireIf,
  StartIsNonNumeric,
  validate
} from 'store/validators';
import AssetSelector from 'components/form/assetselector/AssetSelector';
import { Asset } from 'store/flowContext';
import styles from './TicketRouterForm.module.scss';
import i18n from 'config/i18n';
import TextInputElement from 'components/form/textinput/TextInputElement';
import TembaSelect from 'temba/TembaSelect';
import { fakePropType } from 'config/ConfigProvider';
import { Topic, User } from 'flowTypes';

export interface TicketRouterFormState extends FormState {
  assignee: FormEntry;
  topic: FormEntry;
  ticketer: FormEntry;
  subject: StringEntry;
  body: StringEntry;
  resultName: StringEntry;
}

export default class TicketRouterForm extends React.Component<
  RouterFormProps,
  TicketRouterFormState
> {
  public static contextTypes = {
    config: fakePropType
  };

  constructor(props: RouterFormProps) {
    super(props);

    // if we only have one ticketer, initialize our form with it
    const ticketers = Object.values(this.props.assetStore.ticketers.items);
    const ticketer = ticketers.length === 1 ? ticketers[0] : null;
    this.state = nodeToState(this.props.nodeSettings, ticketer);

    bindCallbacks(this, {
      include: [/^handle/]
    });
  }
  private handleUpdate(
    keys: {
      assignee?: User;
      topic?: Topic;
      ticketer?: Asset;
      subject?: string;
      body?: string;
      resultName?: string;
    },
    submitting = false
  ): boolean {
    const updates: Partial<TicketRouterFormState> = {};

    if (keys.hasOwnProperty('assignee')) {
      updates.assignee = validate(i18n.t('forms.assignee', 'Assignee'), keys.assignee, [
        shouldRequireIf(submitting)
      ]);
    }

    if (keys.hasOwnProperty('topic')) {
      updates.topic = validate(i18n.t('forms.topic', 'Topic'), keys.topic, [
        shouldRequireIf(submitting)
      ]);
    }

    if (keys.hasOwnProperty('ticketer')) {
      updates.ticketer = validate(i18n.t('forms.ticketer', 'Ticketer'), keys.ticketer, [
        shouldRequireIf(submitting)
      ]);
    }

    if (keys.hasOwnProperty('subject')) {
      updates.subject = validate(i18n.t('forms.subject', 'Subject'), keys.subject, []);
    }

    if (keys.hasOwnProperty('body')) {
      updates.body = validate(i18n.t('forms.body', 'Body'), keys.body, [
        shouldRequireIf(submitting)
      ]);
    }

    if (keys.hasOwnProperty('resultName')) {
      updates.resultName = validate(i18n.t('forms.result_name', 'Result Name'), keys.resultName, [
        shouldRequireIf(submitting)
      ]);
    }

    const updated = mergeForm(this.state, updates);

    // update our form
    this.setState(updated);
    return updated.valid;
  }

  private handleTicketerUpdate(selected: Asset[]): void {
    this.handleUpdate({ ticketer: selected[0] });
  }

  private handleAssigneeUpdate(assignee: User): void {
    this.handleUpdate({ assignee });
  }

  private handleTopicUpdate(topic: Topic): void {
    this.handleUpdate({ topic });
  }

  private handleSubjectUpdate(subject: string, name: string, submitting = false): boolean {
    return this.handleUpdate({ subject }, submitting);
  }

  private handleBodyUpdate(body: string): boolean {
    return this.handleUpdate({ body });
  }

  private handleResultNameUpdate(value: string): void {
    const resultName = validate(i18n.t('forms.result_name', 'Result Name'), value, [
      Required,
      Alphanumeric,
      StartIsNonNumeric
    ]);
    this.setState({
      resultName,
      valid: this.state.valid && !hasErrors(resultName)
    });
  }

  private handleSave(): void {
    // validate all fields in case they haven't interacted
    const valid = this.handleUpdate(
      {
        ticketer: this.state.ticketer.value,
        subject: this.state.subject.value,
        body: this.state.body.value,
        resultName: this.state.resultName.value
      },
      true
    );

    if (valid) {
      this.props.updateRouter(stateToNode(this.props.nodeSettings, this.state));
      this.props.onClose(false);
    }
  }

  private getButtons(): ButtonSet {
    return {
      primary: { name: i18n.t('buttons.ok', 'Ok'), onClick: this.handleSave },
      secondary: {
        name: i18n.t('buttons.cancel', 'Cancel'),
        onClick: () => this.props.onClose(true)
      }
    };
  }

  private renderEdit(): JSX.Element {
    const typeConfig = this.props.typeConfig;

    // if we only have one ticketer or we have issues, show the ticket chooser
    const showTicketers =
      Object.keys(this.props.assetStore.ticketers.items).length > 1 || this.props.issues.length > 0;

    return (
      <Dialog title={typeConfig.name} headerClass={typeConfig.type} buttons={this.getButtons()}>
        <TypeList __className="" initialType={typeConfig} onChange={this.props.onTypeChange} />
        {showTicketers ? (
          <div>
            <p>
              <span>Open ticket via... </span>
            </p>
            <AssetSelector
              key="select_ticketer"
              name={i18n.t('forms.ticketer', 'Ticketer')}
              placeholder="Select the ticketing service to use"
              assets={this.props.assetStore.ticketers}
              onChange={this.handleTicketerUpdate}
              entry={this.state.ticketer}
            />
          </div>
        ) : (
          ''
        )}

        <div style={{ display: 'flex', width: '100%', marginTop: '0.5em' }}>
          <div style={{ flexBasis: 250 }}>
            <TembaSelect
              key="select_topic"
              name={i18n.t('forms.topic', 'Topic')}
              endpoint={this.context.config.endpoints.topics}
              onChange={this.handleTopicUpdate}
              value={this.state.topic.value}
              createPrefix={i18n.t('forms.topic_prefix', 'Create Topic: ')}
              searchable={true}
            />
          </div>

          <div style={{ flexGrow: 1, marginLeft: '0.5em' }}>
            <TembaSelect
              key="select_assignee"
              name={i18n.t('forms.assignee', 'Assignee')}
              placeholder="Assign to (Optional)"
              valueKey="email"
              endpoint={this.context.config.endpoints.users}
              onChange={this.handleAssigneeUpdate}
              clearable={true}
              value={this.state.assignee.value}
              getName={(user: User) => {
                if (!user.first_name && !user.last_name) {
                  return user.email || '';
                }
                return `${user.first_name} ${user.last_name}`;
              }}
            />
          </div>
        </div>
        <div className={styles.body}>
          <TextInputElement
            name={i18n.t('forms.body', 'Body')}
            placeholder={i18n.t('forms.enter_a_body', 'Enter a body')}
            entry={this.state.body}
            onChange={this.handleBodyUpdate}
            autocomplete={true}
            textarea={true}
          />
        </div>

        {createResultNameInput(this.state.resultName, this.handleResultNameUpdate)}
        {renderIssues(this.props)}
      </Dialog>
    );
  }

  public render(): JSX.Element {
    return this.renderEdit();
  }
}
