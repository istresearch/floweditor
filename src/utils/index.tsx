import {
  Action,
  Case,
  Category,
  ContactProperties,
  FlowPosition,
  LocalizationMap
} from 'flowTypes';
import { Query } from 'immutability-helper';
import * as React from 'react';
import Localization, { LocalizedObject } from 'services/Localization';
import { Asset, Assets } from 'store/flowContext';
import { FormEntry } from 'store/nodeEditor';
import { v4 as generateUUID } from 'uuid';
import variables from 'variables.module.scss';

export const V4_UUID = /^[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/i;
const LABEL_CHARS = /^[a-zA-Z0-9-][a-zA-Z0-9- ]*$/;
const MAX_LABEL_LENGTH = 36;
const SNAKED_CHARS = /\s|-+(?=\S)/g;

export const MAX_REFLOW_ATTEMPTS = 100;
export const ACTIVITY_INTERVAL = 5000;
export const COLLISION_FUDGE = 5;
export const GRID_SIZE: number = parseInt(variables.grid_size, 10) || 20;
export const NODE_SPACING: number = parseInt(variables.node_spacing, 10) || 10;
export const NODE_PADDING: number = parseInt(variables.node_padding, 10) || 10;
export const QUIET_NOTE = 2000;
export const CONFIRMATION_TIME = 2000;

export interface BoolMap {
  [key: string]: boolean;
}

interface Bounds {
  left: number;
  top: number;
  right: number;
  bottom: number;
}

export type LabelIdCb = (label?: string, labelKey?: string, valueKey?: string) => string;

/**
 * Adjusts the left and top offsets to a grid
 * @param left horizontal offset
 * @param top vertical offset
 */
export const snapToGrid = (left: number, top: number): { left: number; top: number } => {
  let leftAdjust = left % GRID_SIZE;
  let topAdjust = top % GRID_SIZE;

  if (leftAdjust > GRID_SIZE / 3) {
    leftAdjust = GRID_SIZE - leftAdjust;
  } else {
    leftAdjust = leftAdjust * -1;
  }

  if (topAdjust > GRID_SIZE / 3) {
    topAdjust = GRID_SIZE - topAdjust;
  } else {
    topAdjust = topAdjust * -1;
  }

  return {
    left: Math.max(left + leftAdjust, 0),
    top: Math.max(top + topAdjust, 0)
  };
};

/**
 * Adjusts the position offsets to a grid
 */
export const snapPositionToGrid = (position: FlowPosition): FlowPosition => {
  let leftAdjust = position.left % GRID_SIZE;
  let topAdjust = position.top % GRID_SIZE;

  if (leftAdjust > GRID_SIZE / 3) {
    leftAdjust = GRID_SIZE - leftAdjust;
  } else {
    leftAdjust = leftAdjust * -1;
  }

  if (topAdjust > GRID_SIZE / 3) {
    topAdjust = GRID_SIZE - topAdjust;
  } else {
    topAdjust = topAdjust * -1;
  }

  const left = Math.max(position.left + leftAdjust, 0);
  const top = Math.max(position.top + topAdjust, 0);
  const right = left + position.right - position.left;
  const bottom = top + position.bottom - position.top;

  return {
    left,
    top,
    right,
    bottom
  };
};

/**
 * Turns a string array into a bool map for constant lookup
 * @param {string[]} array - an array of strings, e.g. contact fields
 * @returns {object} A map of each string
 */
export const toBoolMap = (array: string[]): BoolMap =>
  array.reduce(
    (map: BoolMap, item: string) => ({
      ...map,
      [item]: true
    }),
    {}
  );

/**
 * Inserts commas into numbers where appropriate for better readability
 * @param {number} value - A number, e.g. 10000
 * @returns {string} A comma-separated string, e.g. 10,000
 */
export const addCommas = (value: number): string =>
  value.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',');

/**
 * Replaces spaces with underscores
 * @param {string} value - A space-separated string to be snaked, e.g. a raw flow field name ('my flow field')
 * @returns {string} A snaked string, e.g. 'my_flow_field'
 */
export const snakify = (value: string): string =>
  value
    .toLowerCase()
    .trim()
    .replace(SNAKED_CHARS, '_');
/**
 * Returns true if a given UUID matches v4 format
 * @param {string} uuid - A version 4 UUID (no braces, uppercase OK)
 * @returns {boolean}
 */
export const validUUID = (uuid: string): boolean => V4_UUID.test(uuid);

/**
 * Returns a given string in title case, e.g. 'full name' becomes 'Full Name'
 * @param {string} str - string to be title-cased
 * @returns {string} Title-cased string
 */
export const titleCase = (str: string): string =>
  str.replace(/\b\w+/g, s => s.charAt(0).toUpperCase() + s.substr(1).toLowerCase());

export const getSelectClassForEntry = (entry: FormEntry): string => {
  if (entry && entry.validationFailures && entry.validationFailures.length > 0) {
    return 'select-invalid';
  }
  return '';
};

export const getSelectClass = (errors: number): string => {
  if (errors === 0) {
    return 'react-select select-base';
  }
  // We use a global selector here for react-select
  return 'react-select select-base select-invalid';
};

export const reorderList = (list: any[], startIndex: number, endIndex: number): any[] => {
  const [removed] = list.splice(startIndex, 1);

  list.splice(endIndex, 0, removed);

  return list;
};

/**
 * Compares basic objects (no methods and DOM nodes; property order important)
 * @param {object} objA - basic object
 * @param {object} objB - basic object
 * @returns {boolean}
 */
export const jsonEqual = (objA: {}, objB: {}): boolean =>
  JSON.stringify(objA) === JSON.stringify(objB);

/**
 * Checks whether any of a list of error strings contain one or more queries.
 * Used to determine whether a particular error has been encountered.
 */
export const hasErrorType = (errors: string[], exps: RegExp[]): boolean => {
  if (!errors.length) {
    return false;
  }
  for (const error of errors) {
    for (const exp of exps) {
      if (error.match(exp)) {
        return true;
      }
    }
  }
  return false;
};

export const getLocalizedObject = (localizations: LocalizedObject[]) => {
  if (localizations && localizations.length) {
    return localizations[0];
  }
};

export interface ClickHandler {
  // onClick: (event: React.MouseEvent<HTMLDivElement>) => void;
  onMouseDown: (event: React.MouseEvent<HTMLDivElement>) => void;
  onMouseUp: (event: React.MouseEvent<HTMLDivElement>) => void;
}

/**
 * Creates a simple click handler via onMouseDown and onMouseUp.
 * This is a necessity in order to let jsPlumb manage our element dragging.
 * @param onClick
 */
export const createClickHandler = (
  onClick: (event: React.MouseEvent<HTMLElement>) => void,
  shouldCancelClick: () => boolean = null,
  onMouseDown?: (event: React.MouseEvent<HTMLElement>) => void
): any => {
  return {
    onMouseDown: (event: React.MouseEvent<HTMLElement>) => {
      if (onMouseDown) {
        onMouseDown(event);
      }
    },
    onMouseUp: (event: React.MouseEvent<HTMLElement>) => {
      if (!shouldCancelClick || !shouldCancelClick()) {
        onClick(event);
      }
    }
  };
};

export const getLocalization = (
  obj: Action | Category | Case,
  localization: LocalizationMap,
  language: Asset
) => Localization.translate(obj, language, localization[language.id]);

/** istanbul ignore next */
export const dump = (thing: any) => console.log(JSON.stringify(thing, null, 4));

/**
 * Apply emphasis style
 */
export const emphasize = (text: string) => <span className="emphasize">{text}</span>;
export const ellipsize = (text: string, max: number = 250) => {
  if (text.length > max) {
    return text.substr(0, max) + '...';
  }
  return text;
};

/**
 * Does property exist in our ContactProperties enum?
 * @param {string} propertyToCheck - 'name' key on react-select AutoCompleteOption
 */
export const propertyExists = (propertyToCheck: string) => {
  for (const property of Object.keys(ContactProperties)) {
    if (property.toLowerCase() === propertyToCheck.toLowerCase().trim()) {
      return true;
    }
  }
  return false;
};

/**
 * Should x element be rendered?
 */
export const renderIf = (predicate: boolean) => (then: JSX.Element, otherwise?: JSX.Element) =>
  predicate ? then : otherwise ? otherwise : null;

/**
 * Does the label meet our length requirements?
 * @param {string} label - label created by react-select
 */
export const properLabelLength = (label: string = '') =>
  label.length > 0 && label.length <= MAX_LABEL_LENGTH;

/**
 * Does the label meet our character requirements?
 * @param {string} label
 */
export const containsOnlyLabelChars = (label: string = '') => LABEL_CHARS.test(label);

/**
 * Does the label meet our length, character requirements?
 * @param {string} label - label created by react-select
 */
export const isValidLabel = (label: string) =>
  properLabelLength(label) && containsOnlyLabelChars(label);

export const isRealValue = (obj: any) => obj !== null && obj !== undefined;

/* istanbul ignore next */
export const timeStart = (name: string) =>
  process.env.NODE_ENV === 'development' && console.time(name);

/* istanbul ignore next */
export const timeEnd = (name: string) =>
  process.env.NODE_ENV === 'development' && console.timeEnd(name);

export const log = (...message: any[]) =>
  process.env.NODE_ENV === 'development' && console.log(...message);

export const capitalize = (str: string) =>
  str.replace(/(?:^|\s)\S/g, captured => captured.toUpperCase());

export const set = (val: any): Query<any> => ({ $set: val });

export const setTrue = (): Query<true> => set(true);

export const setFalse = (): Query<false> => set(false);

export const setNull = (): Query<null> => set(null);

export const setEmpty = (): Query<string> => set('');

export const merge = (val: any): Query<any> => ({ $merge: val });

export const unset = (val: any): Query<any> => ({ $unset: val });

export const push = (arr: any[]): Query<any[]> => ({ $push: arr });

// tslint:disable-next-line:array-type
export const splice = (arr: Array<Array<any>>): Query<Array<Array<any>>> => ({
  $splice: arr
});

export const optionExists = (newOptName: string, options: any[]) =>
  options.find(({ name }) => name.toLowerCase().trim() === newOptName.toLowerCase().trim())
    ? true
    : false;

/* export const isOptionUnique: IsOptionUniqueHandler = ({ option, options, labelKey, valueKey }) =>
    !propertyExists(option.name) && !optionExists(option.name, options);

export const isValidNewOption: IsValidNewOptionHandler = ({ label }) => isValidLabel(label);
*/
/* export const composeCreateNewOption = ({
    idCb,
    type
}: {
    idCb: LabelIdCb;
    type: AssetType;
}): NewOptionCreatorHandler => ({ label, labelKey, valueKey }) => ({
    id: idCb(label, labelKey, valueKey),
    name: label,
    type,
    isNew: true
});*/

/**
 * Deduplicate values in an array.
 * Pass it a key to deduplicate based on obj key.
 */
export const uniqueBy = (a: any[], key: string): any[] => {
  const seen: any = {};
  return a.filter((item: any) => {
    const k = item[key];
    return seen.hasOwnProperty(k) ? false : (seen[k] = true);
  });
};

export const downloadJSON = (obj: any, name: string): void => {
  const dataStr =
    'data:text/json;charset=utf-8,' + encodeURIComponent(JSON.stringify(obj, null, 2));
  const downloadAnchorNode = document.createElement('a');
  downloadAnchorNode.setAttribute('href', dataStr);
  downloadAnchorNode.setAttribute('download', name + '.json');
  document.body.appendChild(downloadAnchorNode); // required for firefox
  downloadAnchorNode.click();
  downloadAnchorNode.remove();
};

export const fetchAsset = (asset: Asset) => (assets: Assets, id: string): Promise<Asset> => {
  return new Promise<Asset>((resolve, reject) => {
    resolve(asset);
  });
};

export let createUUID = (): string => {
  return generateUUID();
};

export const seededUUIDs = (seed: number = 1): any => {
  let current = seed;
  const randomNext = () => {
    const x = Math.sin(current++) * 10000;
    return x - Math.floor(x);
  };

  const random = (min: number, max: number): number => {
    return Math.floor(randomNext() * (max - min) + min);
  };

  return () => {
    const randomSeed: number[] = [];
    for (let idx = 0; idx < 16; idx++) {
      randomSeed.push(random(0, 250));
    }
    return generateUUID({ random: randomSeed });
  };
};

export const range = (start: number, end: number) =>
  Array.from({ length: end - start }, (v: number, k: number) => k + start);

export const pluralize = (count: number, noun: string, suffix: string = 's'): string =>
  `${noun}${count !== 1 ? suffix : ''}`;

export const hasString = (names: string[], key: string): boolean =>
  !!names.find((item: string) => item === key);

export const scalarArrayEquals = (a: any[], b: any[]) =>
  a.every((value, index) => value === b[index]);

export const getURNPath = (urn: string) => {
  return urn.split(':')[1];
};

export const copyToClipboard = (text: string) => {
  navigator.clipboard.writeText(text);
};

export const throttle = (func: any, timeout: any) => {
  let ready: boolean = true;

  return (...args: any) => {
    if (!ready) {
      return;
    }

    ready = false;
    func(...args);
    setTimeout(() => {
      ready = true;
    }, timeout);
  };
};

export const traceUpdate = (component: any, prevProps: any, prevState?: any) => {
  const messages: string[] = [];
  Object.entries(component.props).forEach(
    ([key, val]) => prevProps[key] !== val && messages.push(`Prop: '${key}' changed`)
  );
  if (prevState && component.state) {
    Object.entries(component.state).forEach(
      ([key, val]) => prevState[key] !== val && messages.push(`State: '${key}' changed`)
    );
  }

  if (messages.length > 0) {
    log('****  ' + component._reactInternalFiber.type.name + ' changed');
    messages.forEach((message: string) => {
      log(message);
    });
  }
};

export const debounce = (fn: any, quiet: number, closure: any = null) => {
  if (fn.timeout) {
    window.clearTimeout(fn.timeout);
  }
  fn.timeout = window.setTimeout(closure || fn, quiet);
};

export const onNextRender = (fn: any) => {
  window.setTimeout(fn, 0);
};

export const desnake = (text: string): string => {
  if (!text) {
    return text;
  }
  return text
    .split('_')
    .join(' ')
    .toLowerCase();
};

export const bool = (prop: boolean) => {
  return prop ? 'true' : null;
};
