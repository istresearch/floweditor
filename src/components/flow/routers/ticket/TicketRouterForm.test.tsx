import { mock } from 'testUtils';
import { createOpenTicketNode, getRouterFormProps } from 'testUtils/assetCreators';
import { Types } from 'config/interfaces';
import { RenderNode } from 'store/flowContext';
import TicketRouterForm from './TicketRouterForm';
import * as utils from 'utils';
import * as React from 'react';
import { render, fireEvent, fireChangeText, fireTembaSelect } from 'test/utils';

mock(utils, 'createUUID', utils.seededUUIDs());

const ticketForm = getRouterFormProps({
  node: createOpenTicketNode('Need help', 'Where are my cookies'),
  ui: { type: Types.split_by_ticket }
} as RenderNode);

describe(TicketRouterForm.name, () => {
  describe('render', () => {
    it('should render', () => {
      const { baseElement } = render(<TicketRouterForm {...ticketForm} />);
      expect(baseElement).toMatchSnapshot();
    });
  });

  describe('updates', () => {
    it('should save changes', () => {
      const { baseElement, getByText, getAllByTestId, getByTestId, getByLabelText } = render(
        <TicketRouterForm {...ticketForm} />
      );
      expect(baseElement).toMatchSnapshot();

      const okButton = getByText('Ok');
      const subject = getByTestId('Subject');
      const resultName = getByTestId('Result Name');

      // our ticketer, subject, body and result name are required
      fireChangeText(subject, '');
      fireChangeText(resultName, '');
      fireEvent.click(okButton);
      expect(ticketForm.updateRouter).not.toBeCalled();

      // set our subject and result name
      fireChangeText(subject, 'Need help');
      fireChangeText(resultName, 'My Ticket Result');

      fireEvent.click(okButton);
      expect(ticketForm.updateRouter).toBeCalled();
      expect(ticketForm.updateRouter).toMatchCallSnapshot();
    });
  });
});
