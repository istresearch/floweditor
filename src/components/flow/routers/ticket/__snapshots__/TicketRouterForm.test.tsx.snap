// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TicketRouterForm render should render 1`] = `
<body>
  <div>
    <div
      class="dialog"
    >
      <div
        class="header open_ticket"
      >
        <div
          class="title_container"
        >
          <div
            class="title"
          >
            Open Ticket
          </div>
          <div
            class="subtitle"
          />
        </div>
      </div>
      <div
        class="content"
      >
        <div
          class=" type_list"
        >
          <p>
            When a contact arrives at this point in your flow...
          </p>
          <div>
            <div
              class="normal "
            >
              <temba-select
                data-testid="temba_select_type"
                errors="[]"
                expressions=""
                name="type"
                namekey="description"
                searchable="true"
                valuekey="type"
                values="[{\\"type\\":\\"open_ticket\\",\\"name\\":\\"Open Ticket\\",\\"description\\":\\"Open a ticket with a human agent\\",\\"localizeableKeys\\":[\\"exits\\"],\\"aliases\\":[\\"split_by_ticket\\"],\\"visibility\\":[\\"messaging\\",\\"messaging_background\\",\\"voice\\"],\\"filter\\":\\"ticketer\\"}]"
              />
            </div>
          </div>
        </div>
        
        <div
          class="subject"
        >
          <div
            class="ele"
          >
            <div
              class="wrapper normal"
            >
              <input
                data-testid="Subject"
                name="Subject"
                placeholder="Enter a subject"
                value="Need help"
              />
            </div>
          </div>
        </div>
        <div
          class="body"
        >
          <div
            class="ele"
          >
            <div
              class="wrapper normal"
            >
              <input
                data-testid="Body"
                name="Body"
                placeholder="Enter a body"
                value="Where are my cookies"
              />
            </div>
          </div>
        </div>
        <div
          class="optional_text_input"
        >
          <div
            class="ele"
          >
            <div
              class="label"
            >
              Result Name
            </div>
            <div
              class="wrapper normal"
            >
              <input
                data-testid="Result Name"
                name="Result Name"
                value="Result"
              />
            </div>
            <div
              class="help_text"
            >
              By naming the result, you can reference it later using @results.result
               
            </div>
          </div>
        </div>
      </div>
      <div
        class="footer"
      >
        <div
          class="buttons"
        >
          <div
            class="right_buttons"
          >
            <div
              class="btn secondary  "
              style="margin-left: 0px; margin-top: 0px;"
            >
              Cancel
            </div>
            <div
              class="btn primary  "
              style="margin-left: 10px; margin-top: 0px;"
            >
              Ok
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`TicketRouterForm updates should save changes 1`] = `
<body>
  <div>
    <div
      class="dialog"
    >
      <div
        class="header open_ticket"
      >
        <div
          class="title_container"
        >
          <div
            class="title"
          >
            Open Ticket
          </div>
          <div
            class="subtitle"
          />
        </div>
      </div>
      <div
        class="content"
      >
        <div
          class=" type_list"
        >
          <p>
            When a contact arrives at this point in your flow...
          </p>
          <div>
            <div
              class="normal "
            >
              <temba-select
                data-testid="temba_select_type"
                errors="[]"
                expressions=""
                name="type"
                namekey="description"
                searchable="true"
                valuekey="type"
                values="[{\\"type\\":\\"open_ticket\\",\\"name\\":\\"Open Ticket\\",\\"description\\":\\"Open a ticket with a human agent\\",\\"localizeableKeys\\":[\\"exits\\"],\\"aliases\\":[\\"split_by_ticket\\"],\\"visibility\\":[\\"messaging\\",\\"messaging_background\\",\\"voice\\"],\\"filter\\":\\"ticketer\\"}]"
              />
            </div>
          </div>
        </div>
        
        <div
          class="subject"
        >
          <div
            class="ele"
          >
            <div
              class="wrapper normal"
            >
              <input
                data-testid="Subject"
                name="Subject"
                placeholder="Enter a subject"
                value="Need help"
              />
            </div>
          </div>
        </div>
        <div
          class="body"
        >
          <div
            class="ele"
          >
            <div
              class="wrapper normal"
            >
              <input
                data-testid="Body"
                name="Body"
                placeholder="Enter a body"
                value="Where are my cookies"
              />
            </div>
          </div>
        </div>
        <div
          class="optional_text_input"
        >
          <div
            class="ele"
          >
            <div
              class="label"
            >
              Result Name
            </div>
            <div
              class="wrapper normal"
            >
              <input
                data-testid="Result Name"
                name="Result Name"
                value="Result"
              />
            </div>
            <div
              class="help_text"
            >
              By naming the result, you can reference it later using @results.result
               
            </div>
          </div>
        </div>
      </div>
      <div
        class="footer"
      >
        <div
          class="buttons"
        >
          <div
            class="right_buttons"
          >
            <div
              class="btn secondary  "
              style="margin-left: 0px; margin-top: 0px;"
            >
              Cancel
            </div>
            <div
              class="btn primary  "
              style="margin-left: 10px; margin-top: 0px;"
            >
              Ok
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`TicketRouterForm updates should save changes 2`] = `
Array [
  Object {
    "inboundConnections": Object {},
    "node": Object {
      "actions": Array [
        Object {
          "body": "Where are my cookies",
          "result_name": "My Ticket Result",
          "subject": "Need help",
          "ticketer": Object {
            "name": "Email (bob@acme.com)",
            "uuid": "1165a73a-2ee0-4891-895e-768645194862",
          },
          "type": "open_ticket",
          "uuid": "b1f332f3-bdd3-4891-aec5-1843a712dbf1",
        },
      ],
      "exits": Array [
        Object {
          "destination_uuid": null,
          "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
        },
        Object {
          "destination_uuid": null,
          "uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
        },
      ],
      "router": Object {
        "cases": Array [
          Object {
            "arguments": Array [
              "Success",
            ],
            "category_uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
            "type": "has_category",
            "uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
          },
        ],
        "categories": Array [
          Object {
            "exit_uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
            "name": "Success",
            "uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
          },
          Object {
            "exit_uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
            "name": "Failure",
            "uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
          },
        ],
        "default_category_uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
        "operand": "@results.my_ticket_result",
        "type": "switch",
      },
      "uuid": "132de855-4042-4dc1-a18f-cc2e6a8f790a",
    },
    "ui": Object {
      "config": Object {},
      "position": null,
      "type": "split_by_ticket",
    },
  },
]
`;
